src/storage/txn/scheduler.rs

```diff
@@ -884,52 +881,58 @@ fn process_write_impl(
         }
         Command::ResolveLock {
             ref ctx,
-            start_ts,
-            commit_ts,
+            ref txn_status,
             ref mut scan_key,
-            ref keys,
+            ref key_locks,
         } => {
-            if let Some(cts) = commit_ts {
-                if cts <= start_ts {
-                    return Err(Error::InvalidTxnTso {
-                        start_ts: start_ts,
-                        commit_ts: cts,
-                    });
-                }
-            }
             let mut scan_key = scan_key.take();
-            let mut txn = MvccTxn::new(
-                snapshot,
-                statistics,
-                start_ts,
-                None,
-                ctx.get_isolation_level(),
-                !ctx.get_not_fill_cache(),
-            );
-            let rows = keys.len();
-            for k in keys {
-                match commit_ts {
-                    Some(ts) => txn.commit(k, ts)?,
-                    None => txn.rollback(k)?,
+            let mut modifies: Vec<Modify> = vec![];
+            let mut write_size = 0;
+            let rows = key_locks.len();
+            for &(ref current_key, ref current_lock) in key_locks {
+                let mut txn = MvccTxn::new(
+                    snapshot,
+                    statistics,
+                    current_lock.ts,
+                    None,
+                    ctx.get_isolation_level(),
+                    !ctx.get_not_fill_cache(),
+                );
+                let status = txn_status.get(&current_lock.ts);
+                let commit_ts = match status {
+                    Some(ts) => *ts,
+                    None => panic!("txn status not found!"),
+                };
+                if commit_ts > 0 {
+                    if current_lock.ts >= commit_ts {
+                        return Err(Error::InvalidTxnTso {
+                            start_ts: current_lock.ts,
+                            commit_ts: commit_ts,
+                        });
+                    }
+                    txn.commit(current_key, commit_ts)?;
+                } else {
+                    txn.rollback(current_key)?;
                 }
-                if txn.write_size() >= MAX_TXN_WRITE_SIZE {
-                    scan_key = Some(k.to_owned());
+                write_size += txn.write_size();
+                modifies.append(&mut txn.modifies());
+                if write_size >= MAX_TXN_WRITE_SIZE {
```

Comment:
Please adjust `MAX_TXN_WRITE_SIZE` and `RESOLVE_LOCK_BATCH_SIZE` to make scan_lock never scan duplicated keys.